src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp

```diff
@@ -4875,6 +4898,81 @@ int MethodContext::repGetStringLiteral(CORINFO_MODULE_HANDLE module, unsigned me
     }
 }
 
+void MethodContext::recPrintObjectDescription(void* handle, char* buffer, size_t bufferSize, size_t* pRequiredBufferSize, size_t bytesWritten)
+{
+    if (PrintObjectDescription == nullptr)
+        PrintObjectDescription = new LightWeightMap<DLDL, Agnostic_PrintObjectDescriptionResult>();
+
+    DLDL key;
+    key.A = CastHandle(handle);
+    key.B = (DWORDLONG)bufferSize;
+
+    DWORD strBuf = (DWORD)-1;
+    if (buffer != nullptr && bytesWritten > 0)
+    {
+        size_t bufferRealSize = min(bytesWritten, bufferSize);
+        strBuf = (DWORD)PrintObjectDescription->AddBuffer((unsigned char*)buffer, (DWORD)bufferRealSize);
+    }
+
+    Agnostic_PrintObjectDescriptionResult value;
+    value.bytesWritten = (DWORDLONG)bytesWritten;
+    value.requiredBufferSize = (DWORDLONG)(pRequiredBufferSize == nullptr ? 0 : *pRequiredBufferSize);
+    value.buffer = (DWORD)strBuf;
+
+    PrintObjectDescription->Add(key, value);
+    DEBUG_REC(dmpPrintObjectDescription(key, value));
+}
+void MethodContext::dmpPrintObjectDescription(DLDL key, Agnostic_PrintObjectDescriptionResult value)
+{
+    printf("PrintObjectDescription key hnd-%016llX bufSize-%u, bytesWritten-%u, pRequiredBufferSize-%u", key.A, (unsigned)key.B, (unsigned)value.bytesWritten, (unsigned)value.requiredBufferSize);
+    PrintObjectDescription->Unlock();
+}
+size_t MethodContext::repPrintObjectDescription(void* handle, char* buffer, size_t bufferSize, size_t* pRequiredBufferSize)
+{
+    if (PrintObjectDescription == nullptr)
+    {
+        return 0;
+    }
+
+    DLDL key;
+    key.A = CastHandle(handle);
+    key.B = (DWORDLONG)bufferSize;
+
+    int itemIndex = PrintObjectDescription->GetIndex(key);
+    if (itemIndex < 0)
+    {
+        return 0;
+    }
+    else
+    {
+        Agnostic_PrintObjectDescriptionResult value = PrintObjectDescription->Get(key);
+        DEBUG_REP(dmpPrintObjectDescription(key, value));
+        if (pRequiredBufferSize != nullptr)
+        {
+            *pRequiredBufferSize = (size_t)value.requiredBufferSize;
+        }
```

Comment:
@jakobbotsch I am not sure I understand why it needs such complications on jit side I don't see us doing that for all SPMI APIs we invoke

Reply:
especially for just a poor debug-only api

Reply:
Also, `pRequiredBufferSize` can be null, in fact it is null always currently

Reply:
> @jakobbotsch I am not sure I understand why it needs such complications on jit side I don't see us doing that for all SPMI APIs we invoke

For APIs that we do not care if they fail we are using SPMI error trap on the JIT side normally, e.g. see `eeGetMethodName`, `eeGetFieldName` and `eeTryGetClassSize`. So normally we will make the SPMI implementation fail on missing data and let JIT decide on whether it cares or not. Since I am going to be factoring this code into some common helper function I will probably end up switching it to fail on missing data (in particular we want this behavior for `printMethodName` so we can detect it on the JIT side).

> Also, pRequiredBufferSize can be null, in fact it is always null currently

What I mean is: if `pRequiredBufferSize` is non-null, but the data was not captured during collection. It does not happen today but like the collection/replay to be robust to situations like that.
I understand that this API is only used for diagnostic cases and in those scenarios we don't really care what gets returned here on missing data. It is not going to be the case for `printMethodName` that we need for `JitDisasm`/`JitDump`.